compiler/rustc_hir_typeck/src/mem_categorization.rs

```diff
@@ -82,49 +84,38 @@ impl HirNode for hir::Pat<'_> {
 
 #[derive(Clone)]
 pub(crate) struct MemCategorizationContext<'a, 'tcx> {
-    pub(crate) typeck_results: &'a ty::TypeckResults<'tcx>,
-    infcx: &'a InferCtxt<'tcx>,
-    param_env: ty::ParamEnv<'tcx>,
-    body_owner: LocalDefId,
-    upvars: Option<&'tcx FxIndexMap<HirId, hir::Upvar>>,
+    pub fcx: &'a FnCtxt<'a, 'tcx>,
+    pub upvars: Option<&'tcx FxIndexMap<HirId, hir::Upvar>>,
+    /// The def id of the body that is being categorized.
+    /// This is **different** from `fcx.body_id`.
+    pub inner_body_id: LocalDefId,
```

Comment:
This feels odd, shouldn't we create a separate `FnCtxt` for nested bodies? I guess is issue is that the scoping is as follows?
```
typeck_outer {
    typeck_inner {}

    closure_capture_analysis_outer {}
    closure_capture_analysis_inner {}
}
```
i.e. we already dropped the inner `FnCtxt` again? 🤔

Reply:
The old `FnCtxt` is long gone, though we could create a new one (we don't care about any of the fields in `FnCtxt` except for the body id).

Reply:
All of this is useless though, because we definitely cannot adapt this with the way that clippy is using `ExprUseVisitor`, so I'll have to completely rework this PR.